src/main/kotlin/com/varabyte/kobweb/intellij/colors/KobwebColorProvider.kt

```diff
@@ -0,0 +1,222 @@
+@file:Suppress("UseJBColor")
+
+package com.varabyte.kobweb.intellij.colors
+
+import com.intellij.ide.util.PsiNavigationSupport
+import com.intellij.openapi.editor.ElementColorProvider
+import com.intellij.openapi.fileEditor.OpenFileDescriptor
+import com.intellij.psi.PsiElement
+import com.intellij.psi.impl.source.tree.LeafPsiElement
+import com.intellij.psi.util.PsiUtilBase
+import org.jetbrains.kotlin.analysis.api.analyze
+import org.jetbrains.kotlin.analysis.api.components.KtConstantEvaluationMode
+import org.jetbrains.kotlin.idea.base.psi.kotlinFqName
+import org.jetbrains.kotlin.idea.core.util.toPsiFile
+import org.jetbrains.kotlin.idea.references.mainReference
+import org.jetbrains.kotlin.js.translate.declaration.hasCustomGetter
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.*
+import java.awt.Color
+import kotlin.math.abs
+
+// This prevents the color tracer from checking the codebase ridiculously deep,
+// which might cause lag or even worse a stackoverflow
+private const val MAX_SEARCH_DEPTH = 15
+
+private const val KOBWEB_COLOR_COMPANION_FQ_NAME = "com.varabyte.kobweb.compose.ui.graphics.Color.Companion"
+
+/**
+ * Enables showing those small rectangular color previews next to the line number
+ */
+class KobwebColorProvider : ElementColorProvider {
+
+    override fun getColorFrom(element: PsiElement): Color? = when {
+        element !is LeafPsiElement -> null
+        element.elementType != KtTokens.IDENTIFIER -> null
+        element.parent is KtProperty -> null // Avoid showing multiple previews
+        else -> traceColor(element.parent) // Leaf is just text. The parent is the actual object
+    }
+
+    /**
+     * Setting colors is not (yet) supported
+     */
+    override fun setColorTo(element: PsiElement, color: Color) = Unit
+}
+
+/**
+ * Tries resolving references as deep as possible and checks if a Kobweb color is being referred to
+ */
+private fun traceColor(element: PsiElement, currentDepth: Int = 0): Color? {
+    val nextElement = when (element) {
```

Comment:
Just curious -- you're doing a lot of fancy plugin stuff in here. I'm assuming you had some source code you were borrowing from? (As opposed to just figuring things out via trial and error and code completion to help you out)

If you did, and it is another open source project, could you link to it here? Something like

"The following code was inspired by ...."

Reply:
Not really actually. Sure I must have looked through some other plugins but they weren't really helpful. What was very useful however was the PSI viewer plugin which let's you explore the PSI tree